Modules/Core/Common/include/itkTotalProgressReporter.h

```diff
@@ -0,0 +1,120 @@
+/*=========================================================================
+ *
+ *  Copyright Insight Software Consortium
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkTotalProgressReporter_h
+#define itkTotalProgressReporter_h
+
+#include "itkIntTypes.h"
+#include "itkProcessObject.h"
+
+namespace itk
+{
+/** \class TotalProgressReporter
+ *  \brief A progress reporter for all threads.
+ *
+ * All threads conccurrently contribute on progress to the total number of pixels.
+ *
+ * \ingroup ITKCommon
+ */
+class ITKCommon_EXPORT TotalProgressReporter
+{
+public:
+  ITK_DISALLOW_COPY_AND_ASSIGN(TotalProgressReporter);
+
+
+  TotalProgressReporter(ProcessObject * filter,
+                        SizeValueType   totalNumberOfPixels,
+                        SizeValueType   numberOfUpdates = 100,
```

Comment:
Comparing to PR #1571, this uses the old "numberOfUpdates" while #1571 uses a pixelsPerUpdate. The later in #1571, would cause fewer updates on smaller images or smaller threaded regions. Additionally, the pixelPerUpdate would have a simpler implementation in the `Complete` methods.

Some logic related to timed progress here could be added too. I'm not sure if the timing interval parameter is to get the progress events to occur more frequently, or less.

Thoughts?

Reply:
The time parameter is to make reports more predictable in time. For some images it would be more frequent, for some less frequent.

Reply:
In #1571, the minimum update interval is controlled by the time parameter, and the frequency that the main calling thread calls the `UpdateProgress` methods. So #1571 would suppress duplicate calls UpdateProcess call, but during slow filters the time interval could be would not be any faster. For example, median with a really big neighborhood more frequent events will not occur.

Just some thought on the time interval, there are many ways to handle it.

Reply:
Surely. This was a low implementation effort approach. Another way might be to create a timer which checks the current progress?

The approach in #1571 would work well if filters could predict how slow they would be, and invoke pixel completed appropriately. E.g., median with small radius reports pixel completion every 1000 pixels, while median with very large radius reports every 10 pixels.

Reply:
As opposed to timed events, it would be better to recommend users to "poll" the progress every at their chosen time rather than rely on evens occurring at regular intervals. With 3 different multi-threaders getting consistent behavior from the invoking thread would require a bit of refactoring and rewriting.

Reply:
Polling is impossible in case of Slicer CLI modules - the main executable waits for output to cout with progress information, and then updates the GUI accordingly.

Reply:
There are several different ways Slicer modules are executed. For the CLI modules they can be execute as a sub-process or in a loaded library. I don't know what way the currently work by default. And *impossible* is a narrow(?) word. Certainly, solutions exists with varying amount of work. 

For the SimpleFilters module it is quite do able.

The timed events is a *new* feature request, and is much more burden-some than restoring progress to similar to how it was. Both in terms of performance impact on other users and about of refactoring required in the multi-threaders. IMHO the lack of timed updates has not been a problem until ITKv5, so restoring the behavior should be sufficient. Many good implementation of timed events involve an event queue similar to how GUI behave. That is one reason why timed progress updates are likely better handled by the application rather that ITK, because ITK has no event queue.

While #1571 added the timer, it still did not work smoothly with the pool multi-threader because the invoking thread just waited for the other threads after completion the first work unit. So  regular timed updates did not occurred after the first chunk.

This is a complex issue!

Reply:
Sure, this PR can restore granular progress updates. A possible subsequent PR could tackle update timing.